src/weights.jl

```diff
@@ -2,43 +2,112 @@
 ###### Weight vector #####
 
 if VERSION < v"0.6.0-dev.2123"
-    immutable WeightVec{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
+    abstract AbstractWeights{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
+
+    immutable Weights{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Real
+    end
+
+    immutable FrequencyWeights{S<:Integer, T<:Integer, V<:IntegerVector} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Int
+    end
+
+    immutable ProbabilityWeights{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
         values::V
         sum::S
+        bias::Real
     end
 else
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+    abstract AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+
+    immutable Weights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Real
+    end
+
+    immutable FrequencyWeights{S<:Integer, T<:Integer, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
         values::V
         sum::S
+        bias::Int
+    end
+
+    immutable ProbabilityWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Real
     end
 end
 
 """
-    WeightVec(vs, [wsum])
+    Weights(vs, [wsum])
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
+Construct a `Weights` with weight values `vs` and sum of weights `wsum`.
 If omitted, `wsum` is computed.
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+function Weights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs); corrected::Bool=true)
+    if isempty(vs) || !corrected
+        return Weights{typeof(s), eltype(vs), V}(vs, s, s)
+    else
+        return Weights{S, eltype(vs), V}(vs, s, s * (1 - sum(normalize(vs, 1) .^ 2)))
+    end
+end
+
+function FrequencyWeights{S<:Integer, V<:IntegerVector}(vs::V, s::S=sum(vs); corrected::Bool=true)
+    return FrequencyWeights{S, eltype(vs), V}(vs, s, s - Int(corrected))
 end
 
+# TODO: constructor for ProbabilityWeights, but I'm not familiar with how bias correction works with these
+# types of weights or if bias correction even makes sense.
+# https://en.wikipedia.org/wiki/Inverse_probability_weighting
+
 """
     weights(vs)
 
-Construct a `WeightVec` from a given array.
+Construct a `Weights` type from a given array.
+"""
+weights(vs::RealVector, corrected=true) = Weights(vs; corrected=corrected)
+weights(vs::RealArray, corrected=true) = Weights(vec(vs); corrected=corrected)
+
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+    frequency(vs)
```

Comment:
`frequency` and `exponential` are really too broad terms to be claimed for weights. We should find more specific names, like `fweights` and `expweights`. One (smallà advantage is that it would mirror Stata (with also `pweights` and `aweights`).